dbms/src/AggregateFunctions/AggregateFunctionAvg.h

```diff
@@ -117,12 +119,18 @@ class AggregateFunctionAvg final
 
     void insertResultInto(ConstAggregateDataPtr __restrict place, IColumn & to, Arena *) const override
     {
-        if constexpr (IsDecimal<TResult>)
+        if constexpr (IsDecimal<CalculateType>)
         {
             ScaleType left_scale = result_scale - scale;
-            TResult result = this->data(place).sum.value * getScaleMultiplier<TResult>(left_scale)
-                / static_cast<typename TResult::NativeType>(this->data(place).count);
-            static_cast<ColumnDecimal<TResult> &>(to).getData().push_back(result);
+            CalculateType result = this->data(place).sum.value * getScaleMultiplier<CalculateType>(left_scale)
+                / static_cast<typename CalculateType::NativeType>(this->data(place).count);
+            if constexpr ((std::is_same_v<CalculateType, Decimal<Int256>>
+                           || std::is_same_v<CalculateType, Decimal<Int512>>)&&std::
+                              is_integral_v<typename TResult::NativeType>)
+                static_cast<ColumnDecimal<TResult> &>(to).getData().push_back(
+                    result.template degrade<typename TResult::NativeType>());
+            else
+                static_cast<ColumnDecimal<TResult> &>(to).getData().push_back(result);
```

Comment:
this piece of code is too complex, and i'm not sure the logic is right. I think there should be a simple and clear way to do this.